File: 프로그래머스/네트워크.py
def dfs(start, visited, graph):
    visited[start] = True
    for v in graph[start]:
        if visited[v]:
            continue
        if not visited[v]:
            dfs(v, visited, graph)
    return


def solution(n, computers):
    answer = 0
    graph = [[] for _ in range(n)]
    visited = [False] * n

    # computers를 연결관계 graph로 변경
    j = -1
    for c in computers:
        j += 1
        for i in range(len(c)):
            if c[i] == 1 and i != j:
                graph[j].append(i)

    # 0부터 n-1번까지 탐색
    for i in range(n):
        if not visited[i]:
            answer += 1
            dfs(i, visited, graph)

    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Queue;
import java.util.Set;

class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        boolean[] visited = new boolean[n];

        List<Set<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new HashSet<>());
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == j) {
                    continue;
                }
                if (computers[i][j] == 1) {
                    graph.get(i).add(j);
                    graph.get(j).add(i);
                }
            }
        }

        Queue<Integer> queue = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            if (visited[i]) {
                continue;
            }
            answer++;
            queue.add(i);
            while (!queue.isEmpty()) {
                int now = queue.poll();
                visited[now] = true;
                for (int node : graph.get(now)) {
                    if (visited[node]) {
                        continue;
                    }
                    visited[node] = true;
                    queue.add(node);
                }
            }
        }

        return answer;
    }
}